src/core/agents.jl

```diff
@@ -57,60 +73,168 @@ mutable struct Person{T} <: AbstractAgent
+            # It is important to evaluate the macro in the module that it was called at
+            Core.eval($(__module__), expr)
+        end
+    end
 end
+# TODO: I do not know how to merge these two macros to remove code duplication.
+# There should be away that only the 4-argument version is used
+# and the 3-argument version just passes `AbstractAgent` to the 4-argument.
+macro agent(new_name, base_type, super_type, extra_fields)
+    # This macro was generated with the guidance of @rdeits on Discourse:
+    # https://discourse.julialang.org/t/
+    # metaprogramming-obtain-actual-type-from-symbol-for-field-inheritance/84912
+
+    # We start with a quote. All macros return a quote to be evaluated
+    quote
+        let
+            # Here we collect the field names and types from the base type
+            # Because the base type already exists, we escape the symbols to obtain it
+            base_fieldnames = fieldnames($(esc(base_type)))
+            base_fieldtypes = [t for t in getproperty($(esc(base_type)), :types)]
+            base_fields = [:($f::$T) for (f, T) in zip(base_fieldnames, base_fieldtypes)]
+            # Then, we prime the additional name and fields into QuoteNodes
+            # We have to do this to be able to interpolate them into an inner quote.
+            name = $(QuoteNode(new_name))
+            additional_fields = $(QuoteNode(extra_fields.args))
+            # Now we start an inner quote. This is because our macro needs to call `eval`
+            # However, this should never happen inside the main body of a macro
+            # There are several reasons for that, see the cited discussion at the top
+            expr = quote
+                # Also notice that we escape supertype and interpolate it twice
+                # because this is expected to already be defined in the calling module
+                mutable struct $name <: $$(esc(super_type))
+                    $(base_fields...)
+                    $(additional_fields...)
+                end
+            end
+            # @show expr # uncomment this to see that the final expression looks as desired
+            # It is important to evaluate the macro in the module that it was called at
+            Core.eval($(__module__), expr)
+        end
+    end
+end
+
 
 """
-    OSMAgent
-Combine with [`@agent`](@ref) to create an agent type for [`OpenStreetMapSpace`](@ref).
-It attributes the fields
-`id::Int, pos::Tuple{Int,Int,Float64}`
-to the start of the agent type.
+    NoSpaceAgent <: AbstractAgent
+The minimal agent struct for usage with `nothing` as space (i.e., no space).
+It has the field `id::Int`, and potentially other internal fields that
+are not documentated as part of the public API. See also [`@agent`](@ref).
 """
-mutable struct OSMAgent <: AbstractAgent
+mutable struct NoSpaceAgent <: AbstractAgent
     id::Int
-    pos::Tuple{Int,Int,Float64}
-end
+end
```

Comment:
I've tried to create a new agent via

```julia
@agent MyAgent NoSpaceAgent begin end
```

but it seems the exporting of `NoSpaceAgent` doesn't properly work and

```julia
ERROR: UndefVarError: NoSpaceAgent not defined
```

This is a bit of a weird error because this kind of subtyping works just fine internally, e.g. for `ContinuousAgent`.